src/java.management/share/classes/com/sun/jmx/mbeanserver/WeakIdentityHashMap.java

```diff
@@ -120,9 +120,10 @@ public boolean equals(Object o) {
                 return true;
             if (!(o instanceof IdentityWeakReference<?>))
                 return false;
-            IdentityWeakReference<?> wr = (IdentityWeakReference<?>) o;
-            Object got = get();
-            return (got != null && got == wr.get());
+            @SuppressWarnings("unchecked")
+            IdentityWeakReference<T> wr = (IdentityWeakReference<T>) o;
+            T got = get();
+            return got != null && wr.refersTo(got);
         }
```

Comment:
Here you could pre-screen the get() with:

if (this.refersTo(null) != wr.refersTo(null)) return false;

In case one of the two WeakReference(s) is cleared and the other is not, they are not equal and you don't call this.get() in such case.

Reply:
`expunge` is called at `get`, `put`, and `remove`.   While there is a chance that a referent might be cleared,  I will leave this as is and this patch simply replaces to use `refersTo`.

Reply:
You're right. Where it would matter is in expunge() where the Map keys that get polled of the reference queue are all already cleared and when HashMap.remove(key) is called with such cleared key, it is this cleared key that is the target of key.equals(other) method invocation (and not vice versa), so it doesn't matter if .get() is called on such key as it is already cleared.